src/include/duckdb/execution/index/art/fixed_size_allocator.hpp

```diff
@@ -0,0 +1,132 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/execution/index/art/fixed_size_allocator.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/unordered_set.hpp"
+#include "duckdb/common/constants.hpp"
+#include "duckdb/common/vector.hpp"
+#include "duckdb/common/assert.hpp"
+#include "duckdb/common/types/validity_mask.hpp"
+
+namespace duckdb {
+
+struct BufferEntry {
+	BufferEntry(const data_ptr_t &ptr, const idx_t &allocation_count) : ptr(ptr), allocation_count(allocation_count) {
+	}
+	data_ptr_t ptr;
+	idx_t allocation_count;
+};
+
+//! The FixedSizeAllocator provides pointers to fixed-size sections of pre-allocated memory.
+//! The pointers are of type idx_t, and the leftmost byte must always be zero.
+//! The second to fourth byte store the offset into a buffer, and the last four bytes store the buffer ID,
+//! i.e., a position looks like this [0: empty, 1  - 3: offset, 4 - 7: buffer ID].
+class FixedSizeAllocator {
+public:
+	//! Bitwise AND hex values
+	static constexpr idx_t BUFFER_ID_TO_ZERO = 0xFFFFFFFF00000000;
+	static constexpr idx_t OFFSET_TO_ZERO = 0xFF000000FFFFFFFF;
+	static constexpr idx_t OFFSET_AND_FIRST_BYTE_TO_ZERO = 0x00000000FFFFFFFF;
+	static constexpr idx_t BUFFER_ID_AND_OFFSET_TO_ZERO = 0xFF00000000000000;
+	static constexpr idx_t FIRST_BYTE_TO_ZERO = 0x00FFFFFFFFFFFFFF;
+
+	//! Other constants
+	static constexpr idx_t BUFFER_ALLOCATION_SIZE = Storage::BLOCK_ALLOC_SIZE;
+	static constexpr uint8_t OFFSET_SHIFT = sizeof(uint8_t) * 8 * 4;
+	//! We can vacuum 10% or more of the total memory usage of the allocator
+	static constexpr uint8_t VACUUM_THRESHOLD = 10;
+
+	//! Constants for fast offset calculations in the bitmask
+	static constexpr idx_t BASE[] = {0x00000000FFFFFFFF, 0x0000FFFF, 0x00FF, 0x0F, 0x3, 0x1};
+	static constexpr uint8_t SHIFT[] = {32, 16, 8, 4, 2, 1};
+
+public:
+	explicit FixedSizeAllocator(const idx_t &allocation_size);
```

Comment:
Same here - passing an `idx_t` by reference is a negative from a performance perspective.